exonum/src/messages2/mod.rs

```diff
@@ -0,0 +1,274 @@
+// limitations under the License.
+
+//!
+//! Message represents object received from p2p network.
+//! There next flow between objects:
+//! ```text
+//! +---------+           +---------------+                 +------------+
+//! | Vec<u8> |    ->     | SignedMessage |        ->       |  Protocol  |
+//! +---------+  (verify) +---------------+  (deserialize)  +------------+
+//!                 |                             |
+//!                 V                             V
+//!              (      message dropped if failed       )
+//!
+//! ```
+use hex::{FromHex, ToHex};
+use byteorder::{ByteOrder, LittleEndian};
+use failure::Error;
+
+use std::{borrow::Cow, cmp::PartialEq, fmt, mem, ops::Deref};
+
+use crypto::{hash, CryptoHash, Hash, PublicKey};
+use encoding;
+use storage::StorageValue;
+
+pub(crate) use self::{authorization::SignedMessage, helpers::HexStringRepresentation};
+pub use self::{helpers::BinaryForm, protocol::*};
+
+#[macro_use]
+mod compatibility;
+mod authorization;
+mod helpers;
+mod protocol;
+#[cfg(test)]
+mod tests;
+
+/// Version of the protocol. Different versions are incompatible.
+pub const PROTOCOL_MAJOR_VERSION: u8 = 1;
+
+/// Transaction raw buffer.
+/// This struct is used to transfer transactions in network.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct RawTransaction {
+    service_id: u16,
+    transaction_set: TransactionSetPart,
+}
+
+/// Concrete raw transaction transaction inside `TransactionSet`.
+/// This type used inner inside `transactions!` to transfer some set.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+pub struct TransactionSetPart {
+    message_id: u16,
+    payload: Vec<u8>,
+}
+
+impl TransactionSetPart {
+    /// Creates TransactionSetPart from unchecked raw data.
+    pub fn from_raw_unchecked(message_id: u16, payload: Vec<u8>) -> Self {
+        TransactionSetPart {
+            message_id,
+            payload,
+        }
+    }
+
+    /// Converts `TransactionSetPart` back to raw data.
+    pub fn into_raw_parts(self) -> (u16, Vec<u8>) {
+        (self.message_id, self.payload)
+    }
+}
+
+impl RawTransaction {
+    /// Creates new instance of RawTransaction.
+    pub(in messages) fn new(
+        service_id: u16,
+        transaction_set: TransactionSetPart,
+    ) -> RawTransaction {
+        RawTransaction {
+            service_id,
+            transaction_set,
+        }
+    }
+
+    /// Returns user defined data that should be used for deserialization.
+    pub fn transaction_set(self) -> TransactionSetPart {
+        self.transaction_set
+    }
+
+    /// Returns service_id specified for current transaction.
+    pub fn service_id(&self) -> u16 {
+        self.service_id
+    }
+
+}
+
+impl BinaryForm for RawTransaction {
+    fn serialize(&self) -> Result<Vec<u8>, encoding::Error> {
+        let mut buffer = Vec::new();
+        buffer.resize(2, 0);
```

Comment:
Why not creating `vec![0;2]`?

Reply:
Also magic number is not clear without context.

Reply:
@vitvakatu the reason was unclear allocation rules of `vec!` macro. (If macro allocate exactly N bytes, then every push lead to reallocation)